design-proposals/kubevirt_ephemeral_csi_driver.md

```diff
@@ -0,0 +1,79 @@
+
+
+## User Stories
+
+* As a KubeVirt user I want to be guaranteed that if I have hotplugged volumes in my VM, and the virt-launcher pod terminates unexpectedly, I will not lose any data on my volumes.
+
+## Repos
+kubevirt/kubevirt
+
+# Design
+The idea is to implement just the ephemeral parts of the csi spec. Examples of drivers that do this are [csi-driver-image-populate](https://github.com/kubernetes-csi/csi-driver-image-populator) and [secrets-store-csi-driver](https://github.com/kubernetes-sigs/secrets-store-csi-driver).
+
+To support just the ephemeral portions, we need to implement the appropriate parts of the csi driver, and start the [node-driver-registrar](https://github.com/kubernetes-csi/node-driver-registrar) so we can register the driver with kubernetes. The csi driver will have to run on each node, so will require a daemonset. We could make an additional daemonset, or add additional containers to the virt-handler daemonset. The daemonset will require privileged permissions due to the nature of gRPC socket used to communicate between the different CSI components.
+
+In addition we will have to create the CSI driver object as shown below (I used example naming, we can pick something different if we want)
+
+```yaml
+apiVersion: storage.k8s.io/v1
+kind: CSIDriver
+metadata:
+  name: kubevirt.csi.k8s.io
+  labels:
+    app.kubernetes.io/instance: kubevirt.csi.k8s.io
+    app.kubernetes.io/part-of: kubevirt.io
+    app.kubernetes.io/name: kubevirt.csi.k8s.io
+    app.kubernetes.io/component: csi-driver
+spec:
+  # Supports only ephemeral inline volumes.
+  volumeLifecycleModes:
+  - Ephemeral
+```
+
+One implementation detail to work out, is what to use where to create the volumes. The natural initial implemention would be a directory on the node somewhere, maybe the same directory as where other kubevirt specific configuration is stored?
+
+## Scalability
+
+Since the volumes are ephemeral and will be created on the node the virt-launcher pod is started on, the scaling should be the same as an emptyDir or other ephemeral volumes.
+
+## Update/Rollback Compatibility
+
+Since this is a change in how hotplugged volumes would be mounted inside the virt-launcher pod, there is a potential problem in the virt-handler. The virt-handler uses a specific path format to locate the emptyDir volume used in the virt-launcher. This path is then used to bind mount the volumes with the disk.img files. We would have to update virt-handler to temporarily support both emptyDir and csi ephemeral volumes to allow running VMs to still function properly after upgrading.
+
+## Functional Testing Approach
+
+There is an extensive test suite for hotplugged volumes which should still pass when using the csi ephemeral volumes. In addition we should add some csi ephemeral volume specific tests from the k8s csi suite.
+
+# Implementation Phases
+
```

Comment:
One suggestion here. I'd recommend doing a rapid prototype (ignoring code quality and tests) just to prove out the tech is a viable alternative to empty dir for us before investing too heavily into this.

Reply:
If we want to do a PoC first, I would pick an existing CSI driver that also supports ephemeral volumes and just use that. The easiest that comes to mind is the kubernetes hostpath csi driver, which is basically a mock driver for their e2e tests. It supports all the CSI features so they can test those features including ephemeral volumes. We could modify the teardown of that driver to unmount and use it as the PoC driver.

Reply:
yup, whatever works for you.  And if you feel confident enough that this will work without doign a PoC, that's fine as well.